### PR TITLE
[xpu][test] Port distributed _shard tests cases on Intel GPUs

### PR DESCRIPTION
**Pull request by madhumitha0102**
For https://github.com/pytorch/pytorch/issues/114850, we will port distributed tests to Intel GPU.
We will enable Intel GPU with following methods and  keep the original code styles:

Example: "torch.accelerator.current_accelerator()" to determine the accelerator backend

enabled XPU for the following files:

1. test/distributed/checkpoint/_experimental/test_staging.py
2. test/distributed/checkpoint/e2e/test_e2e_save_and_load.py
3. test/distributed/checkpoint/test_file_system_checkpoint_cpu.py
4. test/distributed/checkpoint/test_state_dict_stager.py
5. test/distributed/checkpoint/test_state_dict_utils.py
6. test/distributed/flight_recorder/test_fr_analysis.py
7. test/distributed/optim/test_zero_redundancy_optimizer.py
8. test/distributed/pipelining/test_backward.py
9. test/distributed/pipelining/test_microbatch.py
10. test/distributed/pipelining/test_schedule_multiproc.py

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @aditew01 @gujinghui @PenghuiCheng @jianyuh @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal @mcarilli @ptrblck @leslie-fang-intel @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98 @fengyuan14 @guangyey

### CODE REVIEW

File: test/distributed/checkpoint/_experimental/test_staging.py [hunk 1]
```diff
@@ -1,13 +1,19 @@
...
 import torch
 from torch.distributed.checkpoint._experimental.staging import (
     CheckpointStagerConfig,
     DefaultStager,
 )
-from torch.testing._internal.common_utils import requires_cuda, run_tests, TestCase
+from torch.testing._internal.common_utils import run_tests, TestCase
+
+
+requires_gpu = unittest.skipUnless(
```
Comment (chuanqi129) [inline test/distributed/checkpoint/_experimental/test_staging.py]: Can we move `requires_gpu` in common_utils directly for other files reuse?

Comment (madhumitha0102) [reply test/distributed/checkpoint/_experimental/test_staging.py]: makes sense. moved it

**Review by chuanqi129 — APPROVED** [review]

File: test/distributed/checkpoint/_experimental/test_staging.py [hunk 2]
```diff
@@ -1,15 +1,14 @@
...
-from unittest import skipIf
 
 import torch
 from torch.distributed.checkpoint._experimental.staging import (
     CheckpointStagerConfig,
     DefaultStager,
 )
 from torch.testing._internal.common_device_type import instantiate_device_type_tests
-from torch.testing._internal.common_utils import run_tests, TEST_ACCELERATOR, TestCase
+from torch.testing._internal.common_utils import requires_gpu, run_tests, TestCase
```
Comment (daisyden) [inline test/distributed/checkpoint/_experimental/test_staging.py]: The semantics changes, requires_gpu() only support xpu and cuda, but TEST_ACCELEARATOR support all the accelerators.

File: test/distributed/checkpoint/e2e/test_e2e_save_and_load.py [hunk 3]
```diff
@@ -512,9 +513,14 @@ def test_overwrite(self):
...
     def backend(self):
-        return "nccl"
+        return (
+            torch.distributed.distributed_c10d.Backend.default_device_backend_map.get(
+                device_type
+            )
+        )
 
     @with_comms
+    @requires_accelerator_dist_backend()
```
Comment (daisyden) [inline test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:523]: semantics chagnes, @requires_accelerator_dist_backend() disabled the cpu test.

File: test/distributed/checkpoint/test_file_system_checkpoint_cpu.py [hunk 4]
```diff
@@ -289,9 +291,19 @@ def get_file_path(self) -> str:
         return paths[0]
 
     def load_tensor(self, tensor: ShardedTensor) -> torch.Tensor:
-        res = torch.zeros(tensor.shape, device="cpu") if dist.get_rank() == 0 else None
-        tensor.gather(out=res)
-        return res
+        backend = dist.get_backend()
+        rank = dist.get_rank()
+
+        if backend == "xccl":
```
Comment (daisyden) [inline test/distributed/checkpoint/test_file_system_checkpoint_cpu.py]: Why we need the if else branches? Can we just following the original code to create a cpu tensor?

File: test/distributed/checkpoint/test_file_system_checkpoint_cpu.py [hunk 5]
```diff
@@ -428,14 +440,18 @@ def test_load_rowwise_to_colwise(self, thread_count) -> None:
...
+        device = f"{device_type}:{dist.get_rank()}"
+        model_to_save = MyShardedModel3(src_spec).to(device)
         model_to_save._register_state_dict_hook(state_dict_hook)
         state_dict_to_save = model_to_save.state_dict()
 
         fs_writer = FileSystemWriter(path=path, thread_count=thread_count)
         save_state_dict(state_dict=state_dict_to_save, storage_writer=fs_writer)
 
-        model_to_load = MyShardedModel3(dst_spec).cuda(dist.get_rank())
+        model_to_load = MyShardedModel3(dst_spec).to(device)
```
Comment (daisyden) [inline test/distributed/checkpoint/test_file_system_checkpoint_cpu.py:444]: suggest to change to: model_to_load = MyShardedModel3(dst_spec).to(dist.get_rank())

File: test/distributed/checkpoint/test_state_dict_stager.py [hunk 6]
```diff
@@ -1220,6 +1221,7 @@ def _verify_sharded_tensor_replication(
             )
 
     @with_comms
+    @requires_accelerator_dist_backend()
```
Comment (daisyden) [inline test/distributed/checkpoint/test_state_dict_stager.py]: Will @requires_accelerator_dist_backend() disable the cpu test?

File: test/distributed/pipelining/test_backward.py [hunk 7]
```diff
@@ -53,11 +49,18 @@ def test_stage_backward(self, device):
 
         torch.testing.assert_close(grad_inputs[0], ref_x.grad)
 
+        rtol, atol = None, None
```
Comment (daisyden) [inline test/distributed/pipelining/test_backward.py:56]: please add comments or issues to explain the reason why xpu has different tolerances.

File: test/distributed/pipelining/test_microbatch.py [hunk 8]
```diff
@@ -327,7 +325,11 @@ def test_chunk_spec(self, device):
         ref = mod(x, y)
         out = pipe(x, y)[0]
 
-        torch.testing.assert_close(out, ref)
+        rtol, atol = None, None
```
Comment (daisyden) [inline test/distributed/pipelining/test_microbatch.py:329]: Also please explain the reason.

File: test/distributed/pipelining/test_schedule_multiproc.py [hunk 9]
```diff
@@ -427,8 +428,6 @@ def test_multi_iter(self, ScheduleClass):
             else:
                 schedule.step()
 
-        dist.barrier(device_ids=[self.rank])
```
Comment (daisyden) [inline test/distributed/pipelining/test_schedule_multiproc.py:542]: Why remove this line?

File: test/distributed/checkpoint/test_state_dict_stager.py (hunk 6 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/test_state_dict_stager.py]: For this one, we want to disable the cpu test because the ```@property
    def backend(self) -> str:
        return f"cpu:gloo,{device_type}:{backend}"``` is defined as so. I think if we do not skip cpu for this test, it will return  ```f"cpu:gloo,cpu:gloo"``` which is not expected and raises an error.  refer to https://github.com/pytorch/pytorch/actions/runs/26477027208/job/77969841943?pr=182199

File: test/distributed/checkpoint/_experimental/test_staging.py (hunk 2 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/_experimental/test_staging.py]: changed it back to TEST_ACCELERATOR

File: test/distributed/checkpoint/e2e/test_e2e_save_and_load.py (hunk 3 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/e2e/test_e2e_save_and_load.py:523]: @daisyden The TestNoCPU class's whole point is to test checkpoint behavior without CPU tensors — it expects an accelerator device to place tensors on. Since , we changed ``` def backend(self): ```  from ```'nccl'``` to ```device_type``` , the CI will run the tests on all device_types including ```'cpu'``` and because of the assert, the test fails. I think @requires_accelerator_dist_backend() is necessary here. Please go through this and let me know if you think otherwise https://github.com/pytorch/pytorch/actions/runs/26477027208/job/77969841978?pr=182199

File: test/distributed/checkpoint/test_file_system_checkpoint_cpu.py (hunk 4 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/test_file_system_checkpoint_cpu.py]: following the original code, removed xccl block

File: test/distributed/pipelining/test_schedule_multiproc.py (hunk 9 above)
Comment (madhumitha0102) [reply test/distributed/pipelining/test_schedule_multiproc.py:542]: added it back

File: test/distributed/checkpoint/test_file_system_checkpoint_cpu.py (hunk 5 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/test_file_system_checkpoint_cpu.py:444]: dist.get_rank() returns an int — e.g. 0, 1, 2. When you call .to(int), it interprets that integer as an accelerator device 
For CPU runners on CI, a call .cpu() is expected instead of a rank. So for CI to pass on cpu, to(device) is better or we skip the test for cpu. Let me know what you think.

File: test/distributed/checkpoint/test_state_dict_stager.py (hunk 6 above)
Comment (daisyden) [reply test/distributed/checkpoint/test_state_dict_stager.py]: This case has intention to test cpu, see line 1239, we could not simply skip cpu backend.

File: test/distributed/checkpoint/test_state_dict_stager.py [hunk 10]
```diff
@@ -1252,6 +1254,7 @@ def test_replication_basic(self):
         stager.close()
 
     @with_comms
+    @requires_accelerator_dist_backend()
```
Comment (daisyden) [inline test/distributed/checkpoint/test_state_dict_stager.py]: This case also have intention to test cpu, we could not skip it for cpu.

File: test/distributed/checkpoint/test_state_dict_stager.py [hunk 11]
```diff
@@ -1293,6 +1296,7 @@ def test_replication_dtensors(self):
...
```
Comment (daisyden) [inline test/distributed/checkpoint/test_state_dict_stager.py]: This case also have intention to test cpu, we could not skip it for cpu.

File: test/distributed/checkpoint/test_state_dict_stager.py [hunk 12]
```diff
@@ -1333,6 +1337,7 @@ def test_replication_sharded_tensors(self):
...
```
Comment (daisyden) [inline test/distributed/checkpoint/test_state_dict_stager.py]: This case also have intention to test cpu, we could not skip it for cpu.

File: test/distributed/checkpoint/test_state_dict_stager.py (hunk 6 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/test_state_dict_stager.py]: Changed the backend property instead of changing the test to only accelerator.

File: test/distributed/checkpoint/test_state_dict_stager.py (hunk 10 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/test_state_dict_stager.py]: Addressed all the comments

File: test/distributed/checkpoint/test_state_dict_stager.py (hunk 11 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/test_state_dict_stager.py]: Addressed all the comments

File: test/distributed/checkpoint/test_state_dict_stager.py (hunk 12 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/test_state_dict_stager.py]: Addressed all the comments

**Review by daisyden — APPROVED** [review]

File: test/distributed/optim/test_zero_redundancy_optimizer.py [hunk 13]
```diff
@@ -939,6 +939,8 @@ def _test_zero_join(self, device):
         NUM_EPOCHS = 2
         LR = 0.01
         torch.manual_seed(0)
+        if device == "xpu":
```
Comment (etaf) [inline test/distributed/optim/test_zero_redundancy_optimizer.py:942]: May need to explain why only xpu need this.

File: test/distributed/checkpoint/_experimental/test_staging.py [hunk 14]
```diff
@@ -148,7 +148,7 @@ def test_different_option_combinations(self) -> None:
     def test_cuda_tensors_staging(self) -> None:
         """Test staging with CUDA tensors."""
         # Create state dict with CUDA tensors
-        device = torch.accelerator.current_accelerator()
+        device = acc.type if (acc := torch.accelerator.current_accelerator()) else "cpu"
```
Comment (liangan1) [inline test/distributed/checkpoint/_experimental/test_staging.py]: Why need to add the cpu test when the acc is not available?

File: test/distributed/optim/test_zero_redundancy_optimizer.py (hunk 13 above)
Comment (madhumitha0102) [reply test/distributed/optim/test_zero_redundancy_optimizer.py:942]: @Chao1Han . I see that this line was added to solve https://github.com/intel/torch-xpu-ops/issues/1870. Can you help answer?

File: test/distributed/checkpoint/_experimental/test_staging.py (hunk 14 above)
Comment (madhumitha0102) [reply test/distributed/checkpoint/_experimental/test_staging.py]: you are right seems redundant. Let me change it back

**Comment by zou3519** [comment]
@claude review this please and put the review inside a GitHub collapsible section

File: test/distributed/optim/test_zero_redundancy_optimizer.py (hunk 13 above)
Comment (Chao1Han) [reply test/distributed/optim/test_zero_redundancy_optimizer.py:942]: Testing found that accuracy discrepancies emerge after multiple rounds of DDP training. Enabling `use_deterministic_algorithms` resolves the issue. As I understand it, `use_deterministic_algorithms` affects the algorithm selection in oneDNN. Since the model consists entirely of linear layers, we only need to enable `use_deterministic_algorithms` for XPU.
